### ISSUE TITLE
Design Sentence Encoder on Top of Morpheme Embeddings

### ISSUE THREAD

**Issue by marctjones** [marctjones/klareco#49]
## Problem

MorphemeAwareEmbedding (Issue #43) produces **word-level** embeddings (176d per word). We need a **sentence-level** encoder for:
1. FAISS index (sentence embeddings for retrieval)
2. Reasoning core input (sentence representations)
3. Semantic similarity tasks

**Current gap**: No sentence encoder using morpheme embeddings

### Option 1: AST-Based Tree Encoder (Recommended)

Use AST structure to aggregate word embeddings hierarchically:

```python
class ASTSentenceEncoder(nn.Module):
    def __init__(self, morpheme_embedder, hidden_dim=256):
        self.morpheme_embedder = morpheme_embedder
        
        # Aggregate AST nodes
        self.tree_lstm = TreeLSTM(
            input_dim=176,  # morpheme embedding dim
            hidden_dim=hidden_dim
        )
    
    def forward(self, sentence_ast):
        # 1. Embed each word using morpheme composition
        word_embeddings = []
        for word_ast in extract_words(sentence_ast):
            morphemes = extract_morphemes(word_ast)
            word_emb = self.morpheme_embedder(morphemes)
            word_embeddings.append(word_emb)
        
        # 2. Aggregate using AST structure
        sentence_emb = self.tree_lstm(
            sentence_ast,
            word_embeddings
        )
        
        return sentence_emb  # 256d
```

**Pro**: Uses syntactic structure, aligns with AST-native approach
**Con**: Requires TreeLSTM implementation (we have this!)

### Option 2: Transformer Encoder

Use standard transformer to aggregate word embeddings:

```python
class TransformerSentenceEncoder(nn.Module):
    def __init__(self, morpheme_embedder, hidden_dim=256, num_layers=4):
        self.morpheme_embedder = morpheme_embedder
        
        self.transformer = nn.TransformerEncoder(
            nn.TransformerEncoderLayer(
                d_model=176,
                nhead=4
            ),
            num_layers=num_layers
        )
        
        self.pooler = nn.Linear(176, hidden_dim)
    
    def forward(self, sentence_ast):
        # 1. Embed words
        word_embeddings = [...]  # Shape: [seq_len, 176]
        
        # 2. Transform
        transformed = self.transformer(word_embeddings)
        
        # 3. Pool (mean or CLS-style)
        sentence_emb = self.pooler(transformed.mean(dim=0))
        
        return sentence_emb  # 256d
```

**Pro**: Standard architecture, well-understood
**Con**: Ignores AST structure, treats words as sequence

### Option 3: Simple Pooling (Baseline)

Just average morpheme embeddings:

```python
class MeanPoolSentenceEncoder(nn.Module):
    def forward(self, sentence_ast):
        word_embeddings = [
            self.morpheme_embedder(extract_morphemes(word))
            for word in extract_words(sentence_ast)
        ]
        return torch.stack(word_embeddings).mean(dim=0)  # 176d
```

**Pro**: No parameters, fast
**Con**: Ignores word order and syntax

### Option 4: Hierarchical (Two-Stage)

First aggregate phrases, then sentence:

```python
class HierarchicalSentenceEncoder(nn.Module):
    def forward(self, sentence_ast):
        # 1. Embed words
        # 2. Aggregate subject, verb, object separately
        subject_emb = aggregate(sentence_ast['subjekto'])
        verb_emb = embed(sentence_ast['verbo'])
        object_emb = aggregate(sentence_ast['objekto'])
        
        # 3. Compose sentence
        sentence_emb = compose(subject_emb, verb_emb, object_emb)
```

**Pro**: Uses sentence structure explicitly
**Con**: More complex, requires careful design

## Recommendation

**Use Option 1: AST-Based Tree Encoder**

Why:
- ✅ Aligns with Klareco philosophy (AST-native)
- ✅ Uses syntactic structure (subject, verb, object)
- ✅ We already have TreeLSTM implementation
- ✅ Can reuse existing training infrastructure
- ✅ Interpretable (can trace how words combine)

### 1. Adapt TreeLSTM for Morpheme Embeddings

```python
# Current TreeLSTM expects word IDs
# New version expects pre-computed word embeddings

class TreeLSTMEncoder(nn.Module):
    def __init__(
        self,
        morpheme_embedder: MorphemeAwareEmbedding,
        hidden_dim: int = 256
    ):
        self.morpheme_embedder = morpheme_embedder
        self.tree_lstm = TreeLSTM(
            input_dim=176,  # morpheme output
            hidden_dim=hidden_dim
        )
```

### 2. Training Objective

Since we're training morpheme embeddings at word level (Issue #45), we need sentence-level training:

**Contrastive learning** (similar to current approach):
- Positive pairs: Sentences from same article
- Negative pairs: Random sentences
- Loss: Contrastive loss on sentence embeddings

**Can reuse** existing training script, just swap embedder!

### 3. Integration Points

- **Retrieval** (Issue #48): Use sentence embeddings for FAISS
- **Reasoning core**: Input sentence embeddings to reasoning layers
- **Evaluation** (Issue #47): Test compositional sentence understanding

## Time Estimate

- Adapt TreeLSTM: 1-2 hours
- Training script updates: 1 hour
- Training (sentence-level): 2-3 hours
- Evaluation: 1 hour
- **Total**: 5-7 hours

## Dependencies

**Blocked by**:
- Issue #45: Morpheme embedding training (must complete first)

**Blocks**:
- Issue #48: FAISS reindexing (needs sentence encoder)
- Reasoning core integration

## Success Criteria

✅ Sentence encoder successfully composes word embeddings
✅ Preserves compositional semantics from morpheme level
✅ Semantic similarity ≥ 0.6 correlation (Tatoeba test)
✅ Retrieval performance equal or better than old TreeLSTM
✅ Handles unseen word combinations in sentences

## Priority

**High (P1)** - Critical path for retrieval and reasoning core

## Related Issues

- Issue #43: MorphemeAwareEmbedding (word-level)
- Issue #44: Training script
- Issue #45: Morpheme training (blocking)
- Issue #47: Evaluation metrics
- Issue #48: FAISS reindexing (blocked)
